### ISSUE TITLE
[no-use-before-define]: typedefs option doesn't work

### ISSUE THREAD

**Issue by Hotell** [typescript-eslint/typescript-eslint#249]
**Repro**

```JSON
{
  "rules": {
    "typescript/no-use-before-define": ["error",{"typedefs": true}]
  }
}
```

```ts
// Should error according to docs, but it wont

let myVar: StringOrNumber = 'hello'
type StringOrNumber = string | number
```

**Expected Result**

Should work according to docs

**Actual Result**

Doesn't work at all

**Additional Info**

Also dunno if this is considered within the implementation, but there is a common pattern in React to get Props/State types from implementation by leveraging `typeof`. So if `typedefs` is set to `false` this should not raise any error:

```ts
type State = typeof initialState
const initialState = { count: 0, who: ''}
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.3.0` |
| `@typescript-eslint/parser`        | `X.Y.Z` |
| `TypeScript`                       | `3.3.3` |
| `ESLint`                           | `5.12` |
| `node`                             | `8.15` |
| `npm`                              | `X.Y.Z` |

**Comment by bradzacher** [comment]
@armano2 has been working on a solution to this (and other scope issues), see #233.

**Comment by armano2** [comment]
@bradzacher my PR is giving us possibility to support types but it will not fix this issue, we have to add support for it to this rule.

types will be stored in separated property (they can't collide with variables)

**Comment by aMarCruz** [comment]
The same error with the same config and this:

```ts
type Props = { foo: typeof Foo };  // <= 'Foo' was using before it was defined

const Foo = () => {};
```

**Comment by bradzacher** [comment]
Merging into #1856
